### ISSUE TITLE
Insert value into CLOB\BLOB column: ORA-01461: can bind a LONG value only for insert into a LONG column 

### ISSUE THREAD

**Issue by ssnickolay** [erlangbureau/jamdb_oracle#80]
Hi! 
First, thank you for the great work that you have done!

In our project, we use blobs for storing JSON configs and faced the problem of saving large values. I was able to simplify the problem to the small and reproducible scenario (I'm using Elixir, but I'm sure that with Erlang will be the same result:

DB setup:

```sql
CREATE TABLE foo(bar clob);
```

Helper module:
```elixir
defmodule Test do
  def call(value) do
    sql = "insert into foo(bar) values (:1)"
    params = [value]
    Ecto.Adapters.SQL.query!(Repo, sql, params)
  end
end
```

And here are different results:
```elixir
test(String.duplicate("a", 4000))
# => %{num_rows: 1, rows: nil}

test(String.duplicate("a", 4001))
# ** (Jamdb.Oracle.Error) 'ORA-01461: can bind a LONG value only for insert into a LONG column\n'

test(String.duplicate("a", 4001) |> to_charlist)
# ** (Jamdb.Oracle.Error) 'ORA-01461: can bind a LONG value only for insert into a LONG column\n'
```

I guess the problem in `jamdb_oracle/src/jamdb_oracle_tns_encoder.erl` and I'm trying to find a solution. Perhaps you will be able to prompt what needs to be fixed. Thanks!

**Comment by vstavskyi** [comment]
Hi!

LOBS are partially realized. I would prefer to use procedures.

Anyway, I changed in jamdb_oracle_tns_encoder.erl CLOB maximum sizes for SELECT (line 300) and INSERT (lines 248, 283) and for one fixed client input charset, if it's not utf8.

Please test curent version.

**Comment by ssnickolay** [comment]
@vstavskyi `CLOB` has been fixed, thanks!
But what about `BLOB`s? I see the same error about `"LONG value only for insert"`.
I guess the reason is that we send both `BLOB`s and `CLOB`s as `TNS_TYPE_VARCHAR`. To be honest, I don’t see how we can distinguish them at the encoder level.

**Comment by ssnickolay** [comment]
Doh, it was a false success :(

The table:
```sql
CREATE TABLE foo(bar clob, baz varchar2(255));
```

This is works:
```elixir
sql = "insert into foo(bar) values (:1)"
params = [String.duplicate("a", 4001)]
IO.inspect Ecto.Adapters.SQL.query!(Repo, sql, params)
```

This one has the same `ORA-01461: can bind a LONG value only for insert into a LONG column`:
```elixir
sql = "insert into foo(bar, baz) values (:1, :2)"
params = [String.duplicate("a", 4001), "hello"]
IO.inspect Ecto.Adapters.SQL.query!(Repo, sql, params)
```

Thus, adding any column to INSERT returns the old error 😞

**Comment by vstavskyi** [comment]
> we cannot use just size of Data?

It's not actual size, but the value in the description of parameter.

Now, with one bind variable, large binary can be inserted or updated (line 285).
With two or more bind variables it's not working. Why? I don't know yet.

```
{"insert into tabl values (:1)", [<<16#E7,16#99,16#BE,...>>]}
```

```
binary = %Ecto.Query.Tagged{value: <<0xE7,0x99,0xBE>>, type: :binary}
Ecto.Adapters.SQL.query(YourApp.Repo, "insert into tabl values (:1)", [binary])
```

**Comment by ssnickolay** [comment]
> With two or more bind variables it's not working. Why? I don't know yet.

I'll try with two vars when BLOB is the last one (as I mentioned in example with CLOB)

**Comment by vstavskyi** [comment]
I think, one large CLOB/BLOB per statement is under rule for Datatype Limits
`Only one LONG column is allowed per table.`

So, one large value and many regular values can exist in query result or bind variables or returning clause.

[Datatype Limits](https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/datatype-limits.html#GUID-963C79C9-9303-49FE-8F2D-C8AAF04D3095)

**Comment by ssnickolay** [comment]
> So, one large value and many regular values can exist in query result or bind variables or returning clause.

As far as I understand, it's about `LONG` type, but we have *LOBs.

I tried to play with your last changes (about `TNS_TYPE_LONGRAW`) and got the same result.

We can insert a batch of parameters **only** if *LOB field is the last.
Here are three tests (I provide raw SQL, but code was executed via `Ecto.Adapters.SQL.query!/3`):

```sql
CREATE TABLE foo(id integer, f_blob clob, f_str varchar2(250));
```

```sql
# Works:
INSERT INTO foo(id, f_str, f_blob) VALUES (:1, :2, :3) [1, "some str", 'payload > 4000']
# => %{num_rows: 1, rows: nil}

# Does not work:
INSERT INTO foo(id, f_blob, f_str) VALUES (:1, :2, :3) [1, 'payload > 4000', "some str"]
# => ORA-01461: can bind a LONG value only for insert into a LONG column

# Does not work:
INSERT INTO foo(f_str, f_blob, id) VALUES (:1, :2, :3) ["some str", 'payload > 4000', 1]
# => ORA-01483: invalid length for DATE or NUMBER bind variable
```

_Exactly the same I got with BLOB\CLOB_

I think the most important thing in these three tests is that the error depends on **what** comes after the big (with Flag 33554432) value. It seems that Oracle does not understand when the *LOB value ends and reads the excess for the next column.
Maybe the problem is in the building payload itself, but I don't know where to look at the Oracle protocol docs(

**Comment by vstavskyi** [comment]
CLOB/BLOB parameter must be binary <<10,11,12,...>> or , if size < 4000, hex encoded list "0A0B0C...."

**Comment by ssnickolay** [comment]
> BLOB parameter must be binary <<10,11,12,...>>, when payload > 4000 or payload =< 4000
or hex encoded list "0A0B0C...." , when payload =< 4000

Yes, I know. I provided examples with CLOB and long charlist. With BLOB it works only if insert one value, as you said above.

Anyway, I guess, I got the problem with fields ordering - https://dbaclass.com/article/max_string_size-parameter-oracle-12c/ . So, CLOB as LONG should be passed as the last arg.

**Comment by ssnickolay** [comment]
@vstavskyi 
TLDR: The workaround with storing LOB separately works. I just think this may not be _the only_ solution.
I played with `odbc` and native Oracle `odbi` adapters; using them I can save LOB with other columns. Also, an order of fields does not affect anything - all examples that I provided above work well.
Maybe these adapters use another type (not LONG)? Or use other protocol options?

**Comment by vstavskyi** [comment]
As I wrote before, LOBS are partially realized.
There is a complicated LOB processing in native Oracle jdbc driver and don't forget about `DBMS_LOB` package.
